internal/provider/tag_definition_resource.go

```diff
@@ -0,0 +1,551 @@
+
+	"github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/booldefault"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/int64default"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/objectplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringdefault"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+)
+
+// Ensure the implementation satisfies the expected interfaces.
+var (
+	_ resource.Resource                   = &tagDefinitionResource{}
+	_ resource.ResourceWithConfigure      = &tagDefinitionResource{}
+	_ resource.ResourceWithValidateConfig = &tagDefinitionResource{}
+	_ resource.ResourceWithImportState    = &tagDefinitionResource{}
+)
+
+var targetKinds = []string{
+	"meshProject",
+	"meshWorkspace",
+	"meshLandingZone",
+	"meshPaymentMethod",
+	"meshBuildingBlockDefinition",
+}
+
+// NewTagDefinitionResource is a helper function to simplify the provider implementation.
+func NewTagDefinitionResource() resource.Resource {
+	return &tagDefinitionResource{}
+}
+
+// tagDefinitionResource is the resource implementation.
+type tagDefinitionResource struct {
+	client *client.MeshStackProviderClient
+}
+
+// Metadata returns the resource type name.
+func (r *tagDefinitionResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_tag_definition"
+}
+
+// Configure adds the provider configured client to the resource.
+func (r *tagDefinitionResource) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	client, ok := req.ProviderData.(*client.MeshStackProviderClient)
+
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected *MeshStackProviderClient, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+
+		return
+	}
+
+	r.client = client
+}
+
+func (r *tagDefinitionResource) ValidateConfig(ctx context.Context, req resource.ValidateConfigRequest, resp *resource.ValidateConfigResponse) {
+	var config tagDefinitionCreate
+
+	diags := req.Config.Get(ctx, &config)
+
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Validate that metadata.name is equal to spec.target_kind.spec.key
```

Comment:
yes that's always the case, so we should just enforce that convention via the provider

Reply:
I had that in mind, but this means it will only have a spec, which can be inconsistent with other objects. I would be in favor of having as close as possible to how we use the API as we do now, and if we plan on switching to a "flat" structure in the provider, then it makes sense to do that imo

Reply:
Implemented it to be set automatically.